File: src/main/java/com/ternbusty/takoyaki/state/State.java
package com.ternbusty.takoyaki.state;

import com.ternbusty.takoyaki.logger.Logger;
import com.ternbusty.takoyaki.util.Json;
import com.ternbusty.takoyaki.util.json.JsonMap;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Map;

public final class State {
    private static final DateTimeFormatter CREATED_FORMAT =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'");

    public String ociVersion;
    public String id;
    public String status;
    public Integer pid;
    public String bundle;
    public Map<String, String> annotations;
    public String created;

    public State() {}

    public State(String ociVersion, String id, ContainerStatus status,
                 Integer pid, String bundle, Map<String, String> annotations, String created) {
        this.ociVersion = ociVersion;
        this.id = id;
        this.status = status.value;
        this.pid = pid;
        this.bundle = bundle;
        this.annotations = annotations;
        this.created = created;
    }

    public ContainerStatus statusEnum() {
        return ContainerStatus.fromString(status);
    }

    public State withStatus(ContainerStatus s) {
        return new State(ociVersion, id, s, pid, bundle, annotations, created);
    }

    public static Path containerDir(String rootPath, String containerId) {
        return Path.of(rootPath, containerId);
    }

    public static Path statePath(String rootPath, String containerId) {
        return containerDir(rootPath, containerId).resolve("state.json");
    }

    public static boolean exists(String rootPath, String containerId) {
        return Files.exists(statePath(rootPath, containerId));
    }

    public static State load(String rootPath, String containerId) throws IOException {
        Path p = statePath(rootPath, containerId);
        Logger.debug("loading state from " + p);
        return Json.readFile(p, State::fromJson);
    }

    public void save(String rootPath) throws IOException {
        Path dir = containerDir(rootPath, id);
        Files.createDirectories(dir);
        Path p = dir.resolve("state.json");
        Logger.debug("saving state to " + p);
        Json.writeFile(p, toJson());
    }

    public static State create(String ociVersion, String containerId,
                               ContainerStatus status, Integer pid, String bundle,
                               Map<String, String> annotations) {
        String created = OffsetDateTime.now(ZoneOffset.UTC).format(CREATED_FORMAT);
        return new State(ociVersion, containerId, status, pid, bundle, annotations, created);
    }

    public State refreshStatus() {
        if (pid != null && isProcessAlive(pid)) return this;
        return statusEnum() == ContainerStatus.STOPPED ? this : withStatus(ContainerStatus.STOPPED);
    }

    private static boolean isProcessAlive(int pid) {
        Path stat = Path.of("/proc", String.valueOf(pid), "stat");
        try {
            String content = Files.readString(stat);
            int lp = content.lastIndexOf(')');
            if (lp < 0 || lp + 2 >= content.length()) return false;
            char st = content.charAt(lp + 2);
            return st != 'Z' && st != 'X';
        } catch (IOException e) {
            return false;
        }
    }

    public static State fromJson(Object node) {
        if (node == null) return null;
        Map<String, Object> o = JsonMap.asObject(node);
        State s = new State();
        s.ociVersion = JsonMap.str(o, "ociVersion");
        s.id = JsonMap.str(o, "id");
        s.status = JsonMap.str(o, "status");
        s.pid = JsonMap.intBoxed(o, "pid");
        s.bundle = JsonMap.str(o, "bundle");
        s.annotations = JsonMap.strMap(o, "annotations");
        s.created = JsonMap.str(o, "created");
        return s;
    }

    public Object toJson() {
        Map<String, Object> o = JsonMap.obj();
        JsonMap.put(o, "ociVersion", ociVersion);
        JsonMap.put(o, "id", id);
        JsonMap.put(o, "status", status);
        JsonMap.put(o, "pid", pid);
        JsonMap.put(o, "bundle", bundle);
        JsonMap.put(o, "annotations", annotations);
        JsonMap.put(o, "created", created);
        return o;
    }
}

File: src/main/java/com/ternbusty/takoyaki/state/State.kt
package com.ternbusty.takoyaki.state

import com.ternbusty.takoyaki.logger.Logger
import com.ternbusty.takoyaki.util.JsonCodec
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

enum class ContainerStatus(val value: String) {
    CREATING("creating"),
    CREATED("created"),
    RUNNING("running"),
    PAUSED("paused"),
    STOPPED("stopped");

    fun canStart(): Boolean = this == CREATED
    fun canKill(): Boolean = this == CREATED || this == RUNNING || this == PAUSED
    fun canDelete(): Boolean = this == STOPPED

    companion object {
        fun fromString(s: String): ContainerStatus =
            entries.firstOrNull { it.value == s }
                ?: throw IllegalArgumentException("Unknown status: $s")
    }
}

@Serializable
data class State(
    val ociVersion: String? = null,
    val id: String? = null,
    val status: String? = null,
    val pid: Int? = null,
    val bundle: String? = null,
    val annotations: Map<String, String>? = null,
    val created: String? = null,
) {
    @Transient
    private var loadedRootPath: String? = null

    constructor(
        ociVersion: String?, id: String?, status: ContainerStatus,
        pid: Int?, bundle: String?, annotations: Map<String, String>?, created: String?
    ) : this(ociVersion, id, status.value, pid, bundle, annotations, created)

    fun statusEnum(): ContainerStatus = ContainerStatus.fromString(status ?: "stopped")

    fun withStatus(s: ContainerStatus): State = copy(status = s.value)

    @Throws(IOException::class)
    fun save(rootPath: String) {
        val cid = id ?: throw IOException("cannot save state without container id")
        val dir = containerDir(rootPath, cid)
        Files.createDirectories(dir)
        val p = dir.resolve("state.json")
        Logger.debug("saving state to $p")
        JsonCodec.saveToFile(p, this)
    }

    fun refreshStatus(): State {
        val currentPid = pid
        if (currentPid != null && isProcessAlive(currentPid)) {
            if (isFrozen()) return withStatus(ContainerStatus.PAUSED)
            return this
        }
        return if (statusEnum() == ContainerStatus.STOPPED) this else withStatus(ContainerStatus.STOPPED)
    }

    private fun isFrozen(): Boolean {
        val rootPath = loadedRootPath ?: return false
        return try {
            val cid = id ?: return false
            val kc = com.ternbusty.takoyaki.config.KontainerConfig.load(rootPath, cid)
            val cgroupPath = kc.cgroupPath ?: return false
            val freeze = com.ternbusty.takoyaki.cgroup.Cgroup.dir(cgroupPath)
                .resolve("cgroup.freeze")
            if (!Files.exists(freeze)) return false
            "1" == Files.readString(freeze).trim()
        } catch (e: Exception) {
            false
        }
    }

    companion object {
        private val CREATED_FORMAT: DateTimeFormatter =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")

        fun containerDir(rootPath: String, containerId: String): Path =
            Path.of(rootPath, containerId)

        fun statePath(rootPath: String, containerId: String): Path =
            containerDir(rootPath, containerId).resolve("state.json")

        fun exists(rootPath: String, containerId: String): Boolean =
            Files.exists(statePath(rootPath, containerId))

        @Throws(IOException::class)
        fun load(rootPath: String, containerId: String): State {
            val p = statePath(rootPath, containerId)
            Logger.debug("loading state from $p")
            val s = JsonCodec.loadFromFile<State>(p)
            s.loadedRootPath = rootPath
            return s
        }

        fun create(
            ociVersion: String, containerId: String,
            status: ContainerStatus, pid: Int?, bundle: String,
            annotations: Map<String, String>?
        ): State {
            val created = OffsetDateTime.now(ZoneOffset.UTC).format(CREATED_FORMAT)
            return State(ociVersion, containerId, status, pid, bundle, annotations, created)
        }

        private fun isProcessAlive(pid: Int): Boolean {
            val stat = Path.of("/proc", pid.toString(), "stat")
            return try {
                val content = Files.readString(stat)
                val lp = content.lastIndexOf(')')
                if (lp < 0 || lp + 2 >= content.length) return false
                val st = content[lp + 2]
                st != 'Z' && st != 'X'
            } catch (e: IOException) {
                false
            }
        }
    }
}
